config/configtls/configtls.go

```diff
@@ -104,14 +109,26 @@ func (c TLSSetting) loadTLSConfig() (*tls.Config, error) {
 		return nil, fmt.Errorf("for auth via TLS, either both certificate and key must be supplied, or neither")
 	}
 
-	var certificates []tls.Certificate
+	var getCertificate func(*tls.ClientHelloInfo) (*tls.Certificate, error)
 	if c.CertFile != "" && c.KeyFile != "" {
-		var tlsCert tls.Certificate
-		tlsCert, err = tls.LoadX509KeyPair(filepath.Clean(c.CertFile), filepath.Clean(c.KeyFile))
+		lastReload := time.Now()
+		var cert tls.Certificate
+		cert, err = tls.LoadX509KeyPair(c.CertFile, c.KeyFile)
 		if err != nil {
 			return nil, fmt.Errorf("failed to load TLS cert and key: %w", err)
 		}
-		certificates = append(certificates, tlsCert)
+		getCertificate = func(_ *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			now := time.Now()
+			if c.ReloadInterval != 0 && lastReload.Add(c.ReloadInterval).Before(now) {
+				var newCert tls.Certificate
+				newCert, err = tls.LoadX509KeyPair(c.CertFile, c.KeyFile)
+				if err == nil {
+					cert = newCert
+					lastReload = now
```

Comment:
Should there be some form of locking on this function to prevent a flood of attempts to reload the cert when it expires?

Reply:
We don't really consider cert expiry in the code.
Concurrent reloading is possible in a highly trafficked server: all invocations would load the same cert, and have a slightly different `now`.
So the end result of the race would be the same value for `cert` and any of the values of `now`. This is probably still fine functionally.

Maybe a bit of resources can be saved with careful usage of an RWLock, so we only ever reload once.